implementations/rust/ockam/ockam_core/src/error/code.rs

```diff
@@ -0,0 +1,419 @@
+    /// - The exit status returned by a subprocess
+    /// - A numeric error code from some other system
+    /// - Et cetera.
+    ///
+    /// But should generally not be used to hold non-identifying metadata, such
+    /// as the date, device IDs, as that information should be stored on the
+    /// payload itself.
+    ///
+    /// Concretely: two `ErrorCode` with different `extra` values should
+    /// identify types of errors.
+    // TODO: is 32 bits okay on embedded? This puts us to 64 bits for the
+    // structure in practice, but means that we'll alwasy be able to hold OS
+    // errors, as well as `old_error::Error::code`.
+    pub extra: i32,
+}
+
+impl ErrorCode {
+    /// Construct the `ErrorCode` for an error.
+    #[cold]
+    pub fn new(origin: Origin, kind: Kind) -> Self {
+        Self {
+            origin,
+            kind,
+            extra: 0,
+        }
+    }
+    /// Construct the `ErrorCode` for an error which contains an additional
+    /// numeric payload.
+    #[cold]
+    pub fn new_with_extra(origin: Origin, kind: Kind, extra: i32) -> Self {
+        Self {
+            origin,
+            kind,
+            extra,
+        }
+    }
+
+    /// Construct an error code with very little useful information
+    #[cold]
+    pub fn unknown() -> Self {
+        Self {
+            origin: Origin::Unknown,
+            kind: Kind::Unknown,
+            extra: 0,
+        }
+    }
+
+    /// Attach an origin and/or kind to the error, without risk of overwriting more
+    /// precise information value.
+    #[must_use]
+    pub fn update_unknown(
+        mut self,
+        o: impl Into<Option<Origin>>,
+        k: impl Into<Option<Kind>>,
+    ) -> Self {
+        if let (Origin::Unknown, Some(o)) = (self.origin, o.into()) {
+            self.origin = o;
+        }
+        if let (Kind::Unknown, Some(k)) = (self.kind, k.into()) {
+            self.kind = k;
+        }
+        self
+    }
+}
+
+/// Origin indicates the abstract source of an error.
+///
+/// Note that [`Error`](super::Error) should already contain precise origin
+/// information (file, line) where the error originated from.
+///
+// Internal note: Once we stablize the API, we should not remove these, just stop emitting them.
+#[repr(u8)]
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Copy, Serialize, Deserialize)]
+pub enum Origin {
+    /// An error for which there is no way to determine a more specific origin.
+    ///
+    /// Eventually this should also be used for errors which, during
+    /// deserialization, have an unknown `Origin` (for now this is too
+    /// error-prone for various reasons).
+    Unknown = 0,
+    /// Reserved for errors emitted by applications using ockam.
+    Application = 1,
+    /// An error originating from the vault.
+    Vault = 2,
+    /// Errors emitted by the transport layer.
+    Transport = 3,
+    /// Errors from some part of the node implementation — the router or relay,
```

Comment:
There is NBSP symbol after the hyphen o_O As well as on line 259

Reply:
What the ... no idea how that got there but good catch :sweat_smile: